src/main/java/com/ldtteam/structurize/client/BlueprintRenderer.java

```diff
@@ -132,21 +131,45 @@ private void init(final BlockPos anchorPos)
             {
                 final BlockPos blockPos = blockInfo.getPos();
                 BlockState state = blockInfo.getState();
-                if ((state.getBlock() == ModBlocks.blockSubstitution.get() && BlueprintRenderSettings.instance.renderSettings.get(RENDER_PLACEHOLDERS)) ||
-                    state.getBlock() == ModBlocks.blockTagSubstitution.get())
+                if (!BlueprintRenderSettings.instance.renderSettings.get(RENDER_PLACEHOLDERS))
                 {
-                    state = Blocks.AIR.defaultBlockState();
-                }
-                if (state.getBlock() == ModBlocks.blockFluidSubstitution.get())
-                {
-                    state = defaultFluidState;
-                }
-                if (SharedConstants.IS_RUNNING_IN_IDE && serverLevel != null && state.getBlock() == ModBlocks.blockSolidSubstitution.get())
-                {
-                    state = BlockUtils.getWorldgenBlock(serverLevel, anchorPos.offset(blockPos), blueprint.getRawBlockStateFunction().compose(b -> b.subtract(anchorPos)));
-                    if (state == null)
+                    if (state.getBlock() == ModBlocks.blockSubstitution.get())
+                    {
+                        state = Blocks.AIR.defaultBlockState();
+                    }
+                    if (state.getBlock() == ModBlocks.blockTagSubstitution.get())
+                    {
+                        final Optional<BlockEntity> tagTE = tileEntities.stream()
```

Comment:
iirc there is pos to te map in blueprint (tho it's probably compound) if not then create it, searching te list in at worst quadratic time (for each te you search te list again) has no place in rendering

Reply:
Searching a list is at worst a linear time operation.
Sorting it is quadratic which is not what he is doing.

That said the stream is okey here Java 17 has intrinsics on this, but nights idea to use the map is better

Reply:
Yeah, we def don't want to search this in each render tick

Reply:
its in init but still meh, and it gets quadratic since linear search multiplied by (at worst) all TEs in blueprint

Reply:
Searching is not sufficient, it has to remove the old BE from the list and substitute a different one.  I could probably change it to use a map instead of a flat list but that would touch a lot more of the existing code, and maps are slower than lists.  The list of BEs is _usually_ fairly short (although using lots of DO blocks distorts this) so it seemed it would be faster as is, although admittedly I haven't actually measured.

Given that this code is literally reinstantiating all the block entities, a list search is going to be the least of the time consumed, and it's not happening every tick; this is all in a conditional cache.

Reply:
_Usually_ has no place in algorithms "sold" to anyone else
Tag anchor != Tag substitution block, or if that's true then tag substitution block is weirdly named class

Reply:
The in-code name follows the naming convention for the other placeholders.  Its translated name is more descriptive of its intended usage.